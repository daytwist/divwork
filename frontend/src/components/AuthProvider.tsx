import Cookies from "js-cookie";
import {
  createContext,
  Dispatch,
  FC,
  ReactNode,
  SetStateAction,
  useContext,
  useMemo,
  useState,
} from "react";

type Props = {
  children: ReactNode;
};

type AuthContextValue = {
  isSignedIn: boolean;
  setIsSignedIn: Dispatch<SetStateAction<boolean>>;
};

export const AuthContext = createContext<AuthContextValue>(
  {} as AuthContextValue
);

export const AuthProvider: FC<Props> = ({ children }) => {
  const [isSignedIn, setIsSignedIn] = useState<boolean>(false);

  const AuthProviderValue = useMemo(
    () => ({ isSignedIn, setIsSignedIn }),
    [isSignedIn, setIsSignedIn]
  );

  return (
    <AuthContext.Provider value={AuthProviderValue}>
      {children}
    </AuthContext.Provider>
  );
};

export const GetIsSignedIn = () => {
  const { isSignedIn, setIsSignedIn } = useContext(AuthContext);

  if (
    Cookies.get("_access_token") &&
    Cookies.get("_client") &&
    Cookies.get("_uid")
  ) {
    setIsSignedIn(true);
    console.log(isSignedIn);
  } else {
    setIsSignedIn(false);
    console.log(isSignedIn);
  }
};
