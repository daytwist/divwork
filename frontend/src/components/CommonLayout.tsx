import { css, Global } from "@emotion/react";
import { Box, Container, Grid } from "@mui/material";
import { FC } from "react";
import { RouteProps } from "react-router-dom";
import Header from "./Header";

const CommonLayout: FC<RouteProps> = ({ children }) => {
  const global = css`
    * {
      margin: 0;
    }
  `;

  return (
    <div>
      <header>
        <Header />
      </header>
      <main>
        <Global styles={global} />
        <Box m={2} pt={4}>
          <Container maxWidth="lg">
            <Grid container justifyContent="center">
              <Grid item>{children}</Grid>
            </Grid>
          </Container>
        </Box>
      </main>
    </div>
  );
};

export default CommonLayout;
