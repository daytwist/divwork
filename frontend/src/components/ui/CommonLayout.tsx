import { FC } from "react";
import { RouteProps } from "react-router-dom";
import { css, Global } from "@emotion/react";
import {
  Box,
  Container,
  createTheme,
  responsiveFontSizes,
  ThemeProvider,
  Toolbar,
} from "@mui/material";
import Grid2 from "@mui/material/Unstable_Grid2/Grid2";
import { jaJP } from "@mui/x-data-grid";
import MenuBars from "./MenuBars";
import { AlertSnackbar } from "./AlertSnackbar";
import { Footer } from "./Footer";

const CommonLayout: FC<RouteProps> = ({ children }) => {
  let theme = createTheme(
    {
      palette: {
        primary: {
          main: "#1de9b6",
          light: "#6effe8",
          dark: "#00b686",
        },
        secondary: {
          main: "#5a6abf",
          light: "#8d98f2",
          dark: "#23408e",
        },
      },
      breakpoints: {
        values: {
          xs: 0,
          sm: 600,
          md: 768,
          lg: 1025,
          xl: 1536,
        },
      },
    },
    jaJP
  );
  theme = responsiveFontSizes(theme);

  const styles = {
    "*": css`
      margin: 0;
    `,

    main: css`
      display: flex;
      flex-direction: column;
      min-height: calc(103vh - 68.5px - 50px);
    `,

    footer: css`
      margin-top: auto;
    `,
  };

  return (
    <div>
      <ThemeProvider theme={theme}>
        <Global styles={styles} />
        <AlertSnackbar />
        <Box sx={{ display: "flex" }}>
          <MenuBars />
          <Box component="main" sx={{ flexGrow: 1, p: 3 }}>
            <Toolbar />
            <Container sx={{ py: 3 }}>
              <Grid2 container justifyContent="center">
                <Grid2>{children}</Grid2>
              </Grid2>
            </Container>
          </Box>
        </Box>
        <Footer />
      </ThemeProvider>
    </div>
  );
};

export default CommonLayout;
