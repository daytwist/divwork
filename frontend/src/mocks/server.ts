// eslint-disable-next-line import/no-extraneous-dependencies
import { setupWorker } from "msw";
import { handlers } from "./handlers";

export const server = setupWorker(...handlers);
