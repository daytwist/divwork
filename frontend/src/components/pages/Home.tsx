import { FC, useContext } from "react";
import { useNavigate, Link } from "react-router-dom";
import Cookies from "js-cookie";
import { Button, Typography } from "@mui/material";
import Grid2 from "@mui/material/Unstable_Grid2/Grid2";
import { AxiosRequestConfig, AxiosResponse } from "axios";
import { axiosInstance } from "../../utils/axios";
import { AuthResponse } from "../../types";
import { AuthContext } from "../../providers/AuthProvider";
import { SnackbarContext } from "../../providers/SnackbarProvider";

const Home: FC = () => {
  const { setIsSignedIn } = useContext(AuthContext);
  const { handleSetSnackbar } = useContext(SnackbarContext);
  const navigate = useNavigate();

  const handleGuestSignIn = () => {
    const options: AxiosRequestConfig = {
      url: "/auth/guest_sign_in",
      method: "POST",
    };

    axiosInstance(options)
      .then((res: AxiosResponse<AuthResponse>) => {
        console.log(res);
        Cookies.set("_access_token", res.headers["access-token"]);
        Cookies.set("_client", res.headers.client);
        Cookies.set("_uid", res.headers.uid);
        setIsSignedIn(true);
        handleSetSnackbar({
          open: true,
          type: "success",
          message: "ゲストログインしました",
        });
        navigate("/teams", { replace: false });
      })
      .catch((err) => {
        console.log(err);
        handleSetSnackbar({
          open: true,
          type: "error",
          // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
          message: `${err.response.data.errors}`,
        });
      });
  };

  return (
    <div>
      <Typography
        variant="h2"
        component="div"
        gutterBottom
        data-testid="home-title"
      >
        DivWork
      </Typography>
      <Grid2 container direction="column" rowSpacing={2}>
        <Grid2 xs={12}>
          <Button
            variant="contained"
            type="button"
            component={Link}
            to="/sign_up/teams/select"
          >
            ユーザー登録
          </Button>
        </Grid2>
        <Grid2 xs={12}>
          <Button
            color="secondary"
            variant="contained"
            type="button"
            component={Link}
            to="/sign_in"
          >
            ログイン
          </Button>
        </Grid2>
        <Grid2 xs={12}>
          <Button color="secondary" type="button" onClick={handleGuestSignIn}>
            ゲストユーザーでログイン
          </Button>
        </Grid2>
      </Grid2>
    </div>
  );
};

export default Home;
