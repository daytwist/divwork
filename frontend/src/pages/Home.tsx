import { FC, useContext } from "react";
import { useNavigate } from "react-router-dom";
import Cookies from "js-cookie";
import { Button, Container, Grid, Typography } from "@mui/material";
import { AxiosRequestConfig, AxiosResponse } from "axios";
import { axiosInstance } from "../utils/axios";
import { AuthResponse } from "../types";
import { AuthContext } from "../providers/AuthProvider";

const Home: FC = () => {
  const { setIsSignedIn } = useContext(AuthContext);
  const navigate = useNavigate();

  const handleGuestSignIn = () => {
    const options: AxiosRequestConfig = {
      url: "/auth/guest_sign_in",
      method: "POST",
    };

    axiosInstance(options)
      .then((res: AxiosResponse<AuthResponse>) => {
        console.log(res);

        if (res.status === 200) {
          Cookies.set("_access_token", res.headers["access-token"]);
          Cookies.set("_client", res.headers.client);
          Cookies.set("_uid", res.headers.uid);

          setIsSignedIn(true);
          navigate(`/teams/${res.data.data.team_id}`, { replace: false });
        }
      })
      .catch((err) => console.log(err));
  };

  return (
    <Container maxWidth="sm">
      <Typography variant="h2" component="div" gutterBottom>
        DivWork
      </Typography>
      <Grid container direction="column" spacing={2}>
        <Grid item>
          <Button
            variant="contained"
            type="button"
            href="/sign_up/teams/select"
          >
            サインアップ
          </Button>
        </Grid>
        <Grid item>
          <Button variant="outlined" type="button" href="/sign_in">
            サインイン
          </Button>
        </Grid>
        <Grid item>
          <Button color="secondary" type="button" onClick={handleGuestSignIn}>
            ゲストユーザーとしてログイン
          </Button>
        </Grid>
      </Grid>
    </Container>
  );
};

export default Home;
