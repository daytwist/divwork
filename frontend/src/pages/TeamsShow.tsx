import { useState, useEffect, FC } from "react";
import { useParams, Link } from "react-router-dom";
import Cookies from "js-cookie";
import { Avatar, Button, Divider, Grid, Typography } from "@mui/material";
import { AxiosRequestConfig, AxiosResponse } from "axios";
import { axiosInstance } from "../utils/axios";
import { Team, TeamsShowResponse, User } from "../types";
import { TasksBar } from "../components/TasksBar";

const TeamsShow: FC = () => {
  const [team, setTeam] = useState<Team>();
  const [users, setUsers] = useState<User[]>();

  const params = useParams<{ id: string }>();

  const options: AxiosRequestConfig = {
    url: `/teams/${params.id}`,
    method: "GET",
    headers: {
      "access-token": Cookies.get("_access_token") || "",
      client: Cookies.get("_client") || "",
      uid: Cookies.get("_uid") || "",
    },
  };

  useEffect(() => {
    axiosInstance(options)
      .then((res: AxiosResponse<TeamsShowResponse>) => {
        console.log(res);
        setTeam(res.data.team);
        setUsers(res.data.users);
      })
      .catch((err) => {
        console.log(err);
      });
  }, []);

  const totals: number[] = [];
  // eslint-disable-next-line array-callback-return
  users?.map((user) => {
    const total = user.unfinished_tasks_count.reduce(
      (sum: number, element: number) => {
        return sum + element;
      },
      0
    );
    totals.push(total);
  });
  const maxCount: number = Math.max(...totals);

  return (
    <div>
      <Grid container direction="column" spacing={2}>
        <Grid item>
          <Typography
            gutterBottom
            variant="h4"
            component="div"
            data-testid="teams-show-h4"
          >
            {team?.name ? `${team.name}のタスク` : ""}
          </Typography>
        </Grid>
        {users?.map((user) => (
          <Grid item key={user.id}>
            <Grid
              container
              direction="row"
              justifyContent="space-between"
              alignItems="center"
            >
              <Grid item>
                <Grid
                  container
                  direction="column"
                  justifyContent="center"
                  alignItems="center"
                  sx={{
                    width: { xs: 80, sm: 100 },
                    height: { xs: 80, sm: 120 },
                    mr: { xs: 1, sm: 2 },
                  }}
                >
                  {user.avatar ? (
                    <Avatar
                      src={user.avatar}
                      alt="avatar"
                      sx={{
                        width: { sm: 60 },
                        height: { sm: 60 },
                      }}
                    />
                  ) : (
                    <Avatar
                      sx={{
                        width: { sm: 60 },
                        height: { sm: 60 },
                      }}
                    />
                  )}
                  <Button
                    variant="text"
                    size="large"
                    component={Link}
                    to={`/users/${user.id}`}
                    sx={{ p: { xs: 0.5 } }}
                  >
                    {user.name}
                  </Button>
                </Grid>
              </Grid>
              <Grid item>
                <TasksBar user={user} maxCount={maxCount} />
              </Grid>
            </Grid>
            <Divider />
          </Grid>
        ))}
      </Grid>
    </div>
  );
};

export default TeamsShow;
