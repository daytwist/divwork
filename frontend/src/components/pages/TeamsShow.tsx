import { FC, SyntheticEvent, useState } from "react";
import { Link } from "react-router-dom";
import {
  Avatar,
  Divider,
  Typography,
  Stack,
  Button,
  Box,
  Tabs,
  Tab,
} from "@mui/material";
import Grid2 from "@mui/material/Unstable_Grid2/Grid2";
import { PriorityBarChart } from "../models/team/PriorityBarChart";
import { TasksNewButton } from "../models/task/TasksNewButton";
import { LoadingColorRing } from "../ui/LoadingColorRing";
import { useFetchTeam } from "../../hooks/useFetchTeam";
import { TabPanel } from "../ui/TabPanel";
import { DeadlineBarChart } from "../models/team/DeadlineBarChart";

const TeamsShow: FC = () => {
  const { team, users } = useFetchTeam();
  const [tabValue, setTabValue] = useState(0);

  const tabProps = (index: number) => {
    return {
      id: `tab-${index}`,
      "aria-controls": `tabpanel-${index}`,
    };
  };

  const totals: number[] = [];
  // eslint-disable-next-line array-callback-return
  users?.map((user) => {
    const total = user.unfinished_tasks_priority_count.reduce(
      (sum: number, element: number) => {
        return sum + element;
      },
      0
    );
    totals.push(total);
  });
  const maxCount: number = Math.max(...totals);

  const handleSwitchTab = (event: SyntheticEvent, newValue: number) => {
    setTabValue(newValue);
  };

  return (
    <div>
      {team ? (
        <Grid2 container direction="column" rowSpacing={1}>
          <Grid2 xs={12}>
            <Typography
              gutterBottom
              variant="h4"
              component="div"
              data-testid="teams-show-h4"
            >
              {team.name}のタスク
            </Typography>
          </Grid2>
          <Grid2 xs={12}>
            <Stack
              direction="row"
              justifyContent="space-between"
              alignItems="center"
              mb={1}
            >
              <TasksNewButton />
              <Box sx={{ borderBottom: 1, borderColor: "divider" }}>
                <Tabs
                  value={tabValue}
                  onChange={handleSwitchTab}
                  textColor="inherit"
                >
                  <Tab label="優先度" {...tabProps(0)} />
                  <Tab label="納期" {...tabProps(1)} />
                </Tabs>
              </Box>
            </Stack>
          </Grid2>
          {users?.map((user) => (
            <Grid2 xs={12} key={user.id}>
              <Stack
                direction="row"
                justifyContent="space-between"
                alignItems="center"
                mb={1}
              >
                <Link
                  to={`/users/${user.id}`}
                  style={{ textDecoration: "none" }}
                >
                  <Stack
                    component={Button}
                    direction="column"
                    justifyContent="center"
                    alignItems="center"
                    sx={{
                      width: { xs: 80, sm: 100 },
                      height: { xs: 80, sm: 120 },
                      mr: { xs: 0, md: 2 },
                    }}
                  >
                    <Avatar
                      src={user.avatar}
                      alt="avatar"
                      sx={{
                        width: { sm: 60 },
                        height: { sm: 60 },
                      }}
                    />
                    <Typography
                      variant="button"
                      sx={{
                        color: "black",
                        mt: 1,
                      }}
                    >
                      {user.name}
                    </Typography>
                  </Stack>
                </Link>
                <Box
                  sx={{
                    width: { xs: 260, sm: 420, md: 530, lg: 750, xl: 900 },
                  }}
                >
                  <TabPanel value={tabValue} index={0}>
                    <PriorityBarChart user={user} maxCount={maxCount} />
                  </TabPanel>
                  <TabPanel value={tabValue} index={1}>
                    <DeadlineBarChart user={user} maxCount={maxCount} />
                  </TabPanel>
                </Box>
              </Stack>
              <Divider />
            </Grid2>
          ))}
        </Grid2>
      ) : (
        <LoadingColorRing />
      )}
    </div>
  );
};

export default TeamsShow;
