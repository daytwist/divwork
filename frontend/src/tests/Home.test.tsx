/* eslint-disable testing-library/no-unnecessary-act */
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { BrowserRouter, MemoryRouter, Route, Routes } from "react-router-dom";
import { rest } from "msw";
import { Home } from "../components/pages/Home";
import { AuthProvider } from "../providers/AuthProvider";
import { SnackbarProvider } from "../providers/SnackbarProvider";
import { CommonLayout } from "../components/ui/CommonLayout";
import { TeamsShow } from "../components/pages/TeamsShow";
import { server } from "../mocks/server";

describe("Home", () => {
  test("スナップショット", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("ゲストログイン", async () => {
    server.use(
      rest.get("/auth/sessions", (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            is_signed_in: true,
            current_user: {
              email: "guest@example.com",
              uid: "guest@example.com",
              id: 3,
              provider: "email",
              allow_password_change: false,
              name: "ゲスト",
              nickname: null,
              image: null,
              team_id: 1,
            },
          })
        );
      })
    );

    render(
      <MemoryRouter initialEntries={["/"]}>
        <AuthProvider>
          <SnackbarProvider>
            <CommonLayout>
              <Routes>
                <Route path="/" element={<Home />} />
                <Route path="/teams" element={<TeamsShow />} />
              </Routes>
            </CommonLayout>
          </SnackbarProvider>
        </AuthProvider>
      </MemoryRouter>
    );

    // eslint-disable-next-line @typescript-eslint/await-thenable
    await act(() => {
      userEvent.click(
        screen.getByRole("button", { name: "ゲストユーザーでログイン" })
      );
    });

    // チーム一覧ページに遷移する
    expect(await screen.findByTestId("teams-show-h4")).toBeInTheDocument();
    expect(
      await screen.findByText("ゲストログインしました")
    ).toBeInTheDocument();
  });
});
