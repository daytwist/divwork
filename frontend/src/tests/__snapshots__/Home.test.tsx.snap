// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home スナップショット 1`] = `
<div>
  <div
    className="MuiTypography-root MuiTypography-h2 MuiTypography-gutterBottom css-i1hrn3-MuiTypography-root"
  >
    DivWork
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column css-1rl1og4-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <a
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        href="/sign_up/teams/select"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        ユーザー登録
      </a>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <a
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        href="/sign_in"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        ログイン
      </a>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <button
        className="MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1d3f8j8-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        ゲストユーザーでログイン
      </button>
    </div>
  </div>
</div>
`;
